### ISSUE TITLE
Install from source

### ISSUE THREAD

**Issue by Forza-tng** [supakeen/pinnwand#165]
Hi,
Is there a way to install Pinnwand from sources rather than via `pip`? I'd like to be able to customise the html and css of my installation.

Thanks :)

**Comment by supakeen** [comment]
Sure, it should be; download the source tarball from pypi (or github on the [releases](https://github.com/supakeen/pinnwand/releases))-page and then a `pip install .` in that directory.

**Comment by Forza-tng** [comment]
Thanks. This works. It's there a way to make it without pip? I am considering packaging Pinnwand on Alpine and we want to keep dependencies minimal.

**Comment by supakeen** [comment]
I *think* everything you'd need should be in the `pyproject.toml` but I readily admit I'm not the best at pep517. Let me direct a friendly Arch developer into this issue to see what they think!

**Comment by Forza-tng** [comment]
I tried to make an ebuild for Gentoo today, but ran out of steam when I realised i needed to make ebuilds for some of the dependencies as well :/...

So to solve my immediate issue with maintaining an up-to-date Pinnwand, I did a shell script that can create and update a dedicated pinnwand venv from github.

It's available at https://git.tnonline.net/Forza/misc/src/branch/main/pinnwand if anyone is interested.

**Comment by supakeen** [comment]
Which dependencies are missing in Gentoo?

**Comment by Forza-tng** [comment]
> Which dependencies are missing in Gentoo?

Looking in https://github.com/supakeen/pinnwand/blob/master/pyproject.toml we have the following missing dependencies.

From `dependencies`:
* pygments-better-html
* token-bucket
* Possibly sqlalchemy-utc (dev-python/sqlalchemy exists)

From `project.optional-dependencies`:
* pre-commit
* pytest-playwright
* pytest-html

The pdm-pep517 should be supported by Gentoo, though i haven't been able to fully test building Pinnwand due to missing deps. 
https://projects.gentoo.org/python/guide/

**Comment by supakeen** [comment]
Right, you don't need the optional dependencies unless you want to run the full test suite(s) in your packaging pipeline. Generally you can get away with just an import-test (at least that's what's allowed on Fedora).

1. `token-bucket`, would need actual packaging but it's quite straightforward (I maintain the package in Fedora).
2. `sqlalchemy-utc` is a very recent addition, if this causes packagers woes (cc @NeilHanlon) then I can probably vendor it, as it's super small.
3. `pygments-better-html` is also small enough to vendor but I don't want to take away the credits for the solution from @Kwpolska :)

As for `pdm`, I'm as always dissatisfied with the state of package management in Python so I'll gladly take guidance to just switch to `setuptools`'s build system.
